[sw/dif_usbdev] Implement USBDEV DIF Library

alphan opened the pull request:
This change introduces a DIF library for the [USB device](https://docs.opentitan.org/hw/ip/usbdev/doc/).

See also: lowRISC/opentitan#1601 (tracking)

Signed-off-by: Alphan Ulusoy <alphan@google.com>

alphan commented:
I plan to add tests and update the in-tree "driver" to use this library in separate PRs.

alphan commented:
Fixed clang-format issue.

Note: Builds fails on CI because this PR does not include #2093.

mcy commented on sw/device/lib/dif/dif_usbdev.h:
```diff
@@ -0,0 +1,759 @@
+ * @return `kDifUsbdevResultInvalidArgument` if `usbdev` is `NULL`,
+ * `kDifUsbdevResultOK` otherwise.
+ */
+dif_usbdev_result_t dif_usbdev_fill_av_fifo(dif_usbdev_t *usbdev);
```
`av` -> `available`? I'm not a fan of the abbreviation, even if the IP uses it.

mcy commented on sw/device/lib/dif/dif_usbdev.h:
```diff
@@ -0,0 +1,759 @@
+ * empty by calling this function periodically.
+ *
+ * @param usbdev A USB device.
+ * @return `kDifUsbdevResultInvalidArgument` if `usbdev` is `NULL`,
```
Nit: I would tend to suggest something a bit more terse for the return code, personally.

silvestrst replied:
I find it simpler to define an `enum` with error codes, comment all of the variants, and in the function description then simply do:
`@return dif_<ip>_result_t.`

mcy commented on sw/device/lib/dif/dif_usbdev.h:
```diff
@@ -0,0 +1,759 @@
+ * operation.
+ *
+ * @param usbdev A USB device.
+ * @return
+ * - `kDifUsbdevRxPacketInfoInvalidArgument` if `usbdev` or `packet` is `NULL`,
+ * - `kDifUsbdevRxPacketGetInfoResultBusy` if there is already a read buffer
+ * operation in progress,
+ * - `kDifUsbdevRxPacketGetInfoResultNoPackets` if RX FIFO is empty,
+ * - `kDifUsbdevRxPacketGetInfoResultOK` otherwise.
+ */
```
I think this is absolutely the kind of stuff that aught to go in a doc comment on the enum, instead.

mcy commented on sw/device/lib/dif/dif_usbdev.h:
```diff
@@ -0,0 +1,759 @@
+ * entire packet payload is read.
+ */
+dif_usbdev_rx_packet_read_bytes_result_t dif_usbdev_rx_packet_read_bytes(
+    dif_usbdev_t *usbdev, uint8_t *dst, uint32_t dst_len,
```
Should the len parameters be `size_t` instead?

mcy commented on sw/device/lib/dif/dif_usbdev.h:
```diff
@@ -0,0 +1,759 @@
+ * Set the address of a USB device.
+ *
+ * @param usbdev A USB device.
+ * @param addr New address. Only the last 7 bits are significant.
```
> Only the last 7 bits are significant

Why not `uint8_t` then?

alphan replied:
I was on the fence about this because of our word size and other changes that needed to be made to be consistent. Done. Also made changes to fifo depth, endpoint, frame index, etc., but left IRQ-related code as is to be consistent with other DIF libraries.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,934 @@
+} endpoint_hw_info_t;
+
+static const endpoint_hw_info_t kEndpointHwInfos[USBDEV_NUM_ENDPOINTS] = {
+        [0] = {.config_in_reg_offset = USBDEV_CONFIGIN0_REG_OFFSET,
```
Call me crazy, but I feel like a macro to ensure that the numbers are the same in the array index and the two constants would help avoid typos here.

alphan replied:
I agree.

lenary replied:
Let's be careful about going macro-crazy. :)

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,934 @@
+
+static const endpoint_hw_info_t kEndpointHwInfos[USBDEV_NUM_ENDPOINTS] = {
+        [0] = {.config_in_reg_offset = USBDEV_CONFIGIN0_REG_OFFSET,
+               .bit_index = USBDEV_IN_SENT_SENT0},
```
Nit: trailing commas in `{}` initializers helps the formatter reflow code better, so it's not as packed.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,934 @@
+  return irq >= kDifUsbdevIrqFirst && irq <= kDifUsbdevIrqLast;
+}
+
+/**
+ * Static functions for the free buffer pool.
+ */
```
I would have written one comment per helper function. What do you think?

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,934 @@
+    return false;
+  }
+
+  pool->entries[++pool->top] = entry;
```
Maybe split out the `++` into another statement? Most people (myself included) might not remember the precedence of `++` vs `->`.

(In general, I would personally avoid using `++foo` as an expression.)

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,934 @@
+
+static bool buffer_pool_init(dif_usbdev_buffer_pool_t *pool) {
+  // Start with an empty pool
+  pool->top = -1;
```
Consider replacing `-1` with a constant.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,934 @@
+  usbdev->active_buffer_ops = 0;
+}
+
+static void buffer_op_start_write(dif_usbdev_t *usbdev, uint32_t endpoint) {
+  usbdev->active_buffer_ops |= (1 << kEndpointHwInfos[endpoint].bit_index);
+}
+
+static void buffer_op_stop_write(dif_usbdev_t *usbdev, uint32_t endpoint) {
+  usbdev->active_buffer_ops &= ~(1 << kEndpointHwInfos[endpoint].bit_index);
+}
+
+static bool buffer_op_is_writing(dif_usbdev_t *usbdev, uint32_t endpoint) {
+  return (usbdev->active_buffer_ops &
+          (1 << kEndpointHwInfos[endpoint].bit_index)) > 0;
+}
+
+static void buffer_op_start_read(dif_usbdev_t *usbdev) {
+  usbdev->active_buffer_ops |= (1 << USBDEV_READ_BUFFER_OP_BIT_INDEX);
+}
+
+static void buffer_op_stop_read(dif_usbdev_t *usbdev) {
+  usbdev->active_buffer_ops &= ~(1 << USBDEV_READ_BUFFER_OP_BIT_INDEX);
+}
+
+static bool buffer_op_is_reading(dif_usbdev_t *usbdev) {
+  return (usbdev->active_buffer_ops & (1 << USBDEV_READ_BUFFER_OP_BIT_INDEX)) >
+         0;
+}
```
I think I'm eating my words from https://github.com/lowRISC/opentitan/pull/2047, we really aught to have non-MMIO get_bit and set_bit functions. I'll open a CL to add them, and you'll be able to use them here once I'm done.

@silvestrst You were right. =P

alphan replied:
Sure, I don't think we had those when I was writing this. Done.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,934 @@
+  }
+
+  // Check enum fields
+  if (!is_valid_toggle(config->differential_rx) ||
+      !is_valid_toggle(config->differential_tx) ||
+      !is_valid_toggle(config->single_bit_eop) ||
+      !is_valid_power_sense_override(config->power_sense_override)) {
+    return kDifUsbdevResultGeneralError;
```
I think we were only doing this kind of checking in `switch` statements, right? @lenary could you clarify?

alphan replied:
I included this check because an equivalent `switch` would have it but wrote it in this form to make this cheks earlier. @lenary @mcy please let me know if you want me to make any changes.

lenary replied:
From the style guide:
> DIFs must ensure, if they only accept a subset of an enum, that the argument is within that subset. However, DIFs may assume, for checking preconditions, that any enum argument is one of the enum constants.

You're accepting all sates, so you don't have to check `is_valid_toggle`, or `is_valid_power_sense_override`. I'm happy that what you're doing if the values are not one of the enum values is "safe" in the implementation of this function.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,934 @@
+  const uint32_t fifo_entry =
+      mmio_region_read32(usbdev->base_addr, USBDEV_RXFIFO_REG_OFFSET);
+  // Fill packet info
+  packet->endpoint =
+      (fifo_entry >> USBDEV_RXFIFO_EP_OFFSET) & USBDEV_RXFIFO_EP_MASK;
+  packet->is_setup = fifo_entry & USBDEV_RXFIFO_SETUP;
+  packet->len =
+      (fifo_entry >> USBDEV_RXFIFO_SIZE_OFFSET) & USBDEV_RXFIFO_SIZE_MASK;
+  // Get the buffer entry that holds the payload
+  uint32_t buffer_entry =
+      (fifo_entry >> USBDEV_RXFIFO_BUFFER_OFFSET) & USBDEV_RXFIFO_BUFFER_MASK;
+  // Start a read buffer operation
+  buffer_op_start_read(usbdev);
+  usbdev->read_buffer_op_state =
+      (dif_usbdev_read_buffer_op_state_t){.buffer_entry = buffer_entry,
+                                          .remaining_bytes = packet->len,
+                                          .offset = 0};
```
This should all be using `bitfield.h`.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,934 @@
+    bytes_to_copy = dst_len;
+  }
+  // Get the address of the buffer entry
+  uint32_t buffer_addr = USBDEV_BUFFER_REG_OFFSET +
```
Consider factoring this out into a helper?

alphan replied:
Good point.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,934 @@
+  usbdev->read_buffer_op_state.offset += bytes_to_copy;
+  usbdev->read_buffer_op_state.remaining_bytes -= bytes_to_copy;
+  // Check if the entire payload is read
+  bool packet_done = false;
```
Nit: Maybe declare this as the error return type `dif_usbdev_blah_result_t result = kDifWhateverOk;` and override it in the if clause below? What do you think?

alphan replied:
Ah, forgot to clean this up, sorry about that :)

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,934 @@
+  uint32_t buffer_len;
+
+  switch (in_packet_status) {
+    case kDifUsbdevTxPacketStatusNoPacket:
```
Probably a good idea to wrap a big case in braces, just to avoid shenanigans with variable having to declare some variables outside of the `switch`.

alphan replied:
No variables are declared in these `case`s but still this is a good point. Done.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,934 @@
+      return kDifUsbdevInPacketWriteBytesResultBusy;
+    default:
+      return kDifUsbdevInPacketWriteBytesResultGeneralError;
+      break;
```
`break` after `return` is redundant.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,934 @@
+                                   USBDEV_CONFIGIN0_PEND0)) {
+    // Canceled due to an IN SETUP packet
+    // Clear pending bit (rw1c)
+    mmio_region_write32(usbdev->base_addr, config_in_reg_offset,
+                        1 << USBDEV_CONFIGIN0_PEND0);
```
This can be a write-only set bit (your PR to fix my linker snafu got in).

(This applies elsewhere in the file, ofc.)

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,934 @@
+  return kDifUsbdevResultOK;
+}
+
+dif_usbdev_result_t dif_usbdev_irq_disable_all(dif_usbdev_t *usbdev,
+                                               uint32_t *cur_config) {
+  if (usbdev == NULL) {
+    return kDifUsbdevResultInvalidArgument;
+  }
+
+  if (cur_config != NULL) {
+    *cur_config =
+        mmio_region_read32(usbdev->base_addr, USBDEV_INTR_ENABLE_REG_OFFSET);
+  }
+
+  mmio_region_write32(usbdev->base_addr, USBDEV_INTR_ENABLE_REG_OFFSET, 0);
+
+  return kDifUsbdevResultOK;
+}
+
+dif_usbdev_result_t dif_usbdev_irq_restore(dif_usbdev_t *usbdev,
+                                           uint32_t new_config) {
+  if (usbdev == NULL) {
+    return kDifUsbdevResultInvalidArgument;
+  }
+
+  mmio_region_write32(usbdev->base_addr, USBDEV_INTR_ENABLE_REG_OFFSET,
+                      new_config);
+
+  return kDifUsbdevResultOK;
+}
```
This is a really interesting design for touching all of the IRQ bits at once. Nice work!

alphan replied:
Thanks, but the credit goes to @silvestrst :) IIRC, I saw this in the UART DIF library.

mcy commented on sw/device/lib/dif/dif_usbdev.h:
```diff
@@ -0,0 +1,759 @@
+ * `kDifUsbdevResultGeneralError` covers other error cases.
+ */
+typedef enum dif_usbdev_result {
+  kDifUsbdevResultOK,
+  kDifUsbdevResultInvalidArgument,
```
Minor nit: `OK` -> `Ok`, and `InvalidArgument` -> `BadArg` for consistency with other DIFs.

mcy commented on sw/device/lib/dif/dif_usbdev.h:
```diff
@@ -0,0 +1,759 @@
+dif_usbdev_result_t dif_usbdev_interface_enable(dif_usbdev_t *usbdev,
+                                                dif_usbdev_toggle_t new_state);
+
+/**
+ * Information about a received packet.
+ */
+typedef struct dif_usbdev_rx_packet_info {
+  uint32_t endpoint;
+  uint32_t len;
+  bool is_setup;
+} dif_usbdev_rx_packet_info_t;
```
I would suggest that, in structs like this one that the user is intended to inspect, the fields probably want comments, as well.

lenary commented:
I had a more general question that I've talked over with @alphan but we're not 100% sure of the answer.

At the moment, the `dif_usbdev_t` keeps around a single receive buffer, which is initialized with `dif_usbdev_rx_packet_get_info`, and then read with `dif_usbdev_rx_packet_read_bytes` (and `dif_usbdev_rx_packet_discard_bytes`). When you get to the end of that buffer, the buffer is automatically added back to the free buffer pool, and then the DIF can process the next received packet. 

I'm wondering whether this is the right approach, as it only allows one packet to be received and read at a time. The other approach would be to effectively return a structure that represented the received buffer (including the buffer id and other attributes), and then provide specialised functions for reading data out of that buffer. 

The trade-off is that now the receiver has to keep track of the buffer, and return it to the DIF when it is finished with the buffer (though this could maybe still happen automatically depending on the parameters to the function that reads/discards bytes from the buffer). We don't have RAII in C at least, which makes this more error prone.

I think it probably is sensible, if we go down this route, to wrap the receive and transmit buffers up in (separate) struct types, and not to do reading from the buffer pointer directly, in order to gain some more type safety. I am glad we have a slightly higher level of abstraction in the DIF right now, but I worry it may cause issues. @alphan did say he went back and forth on the design a few times, which is why I wanted to raise this discussion in a more general place. 

cc @gkelly

lenary commented on sw/device/lib/dif/dif_usbdev.h:
```diff
@@ -0,0 +1,759 @@
+ * `kDifUsbdevResultGeneralError` if `endpoint` is invalid,
+ * `kDifUsbdevResultOK` otherwise.
+ */
+dif_usbdev_result_t dif_usbdev_clear_data_toggle(dif_usbdev_t *usbdev,
```
This functiontionality is for hw debug only (#2364) so it does not need to be included in the DIF. If needed it can go in a DV-specific extension.

alphan commented on sw/device/lib/dif/meson.build:
```diff
@@ -55,3 +55,15 @@ sw_lib_dif_spi_device = declare_dependency(
+      hw_ip_usbdev_reg_h,
+      'dif_usbdev.c',
+    ],
+    dependencies: [sw_lib_mmio],
```
Done. (It happens to work before this change because `sw_lib_mmio` depends on `sw_lib_bitfield`).

alphan commented on sw/device/lib/dif/dif_usbdev.h:
```diff
@@ -0,0 +1,759 @@
+/**
+ * Hardware constants.
+ */
+#define USBDEV_NUM_ENDPOINTS 12
```
I already have `_Static_assert(USBDEV_NUM_ENDPOINTS == USBDEV_PARAM_NENDPOINTS, ...)` at the top of `dif_usbdev.c`. I defined this macro this way in order not to expose the generated header to all clients.

alphan commented on sw/device/lib/dif/dif_usbdev.h:
```diff
@@ -0,0 +1,759 @@
+ * `kDifUsbdevResultGeneralError` if `endpoint` is invalid,
+ * `kDifUsbdevResultOK` otherwise.
+ */
+dif_usbdev_result_t dif_usbdev_endpoint_stall_get(dif_usbdev_t *usbdev,
+                                                  uint32_t endpoint,
+                                                  bool *state);
```
Yes, from the spec:
> In the case of a protocol stall, the device must send a STALL for all IN/OUT requests until the next SETUP token is received. To support this, software sets the stall register for an endpoint when the host requests an unsupported transfer. The hardware will then send a STALL response to all IN/OUT transactions until the next SETUP is received for this endpoint. Receiving the **SETUP token clears the stall register** for that endpoint. The hardware then sends NACKs to any IN/OUT requests until the software has decided what action to take for the new SETUP request.

Rest of the configuration, i.e. setup, out, and iso, is not modified by the hardware.

alphan commented:
Rebased and addressed review comments.

Note: Pushed changes as a separate commit to give everyone a chance to review the changes. I will squash everything into one commit before merge.

mcy commented on sw/device/lib/dif/dif_usbdev.h:
```diff
@@ -0,0 +1,772 @@
+ */
+typedef struct dif_usbdev_buffer_pool {
+  uint8_t entries[USBDEV_NUM_BUFFERS];
+  int8_t top;
```
Wouldn't it make more sense for this to be named `len`, with empty = 0? It seems like that might make the operations somewhat more natural (I haven't looked at the implementation, though).

alphan replied:
I considered doing that too but thought this was more natural. Implementation-wise I think they are roughly equivalent, i.e. `-1` when checking if the pool is empty/full vs `-1` when adding/removing entries. I don't have a strong opinion on this but as I said I find `top` more natural. Please let me know if you would like me to change this to `len`.

mcy replied:
I don't have a strong feeling either way. Thanks for checking it out.

mcy commented on sw/device/lib/dif/dif_usbdev.h:
```diff
@@ -0,0 +1,772 @@
+ * State of a read buffer operation.
+ *
+ * This struct stores
+ * - the buffer to read from,
+ * - the number of remaining bytes, and
+ * - the offset to start reading from
```
Maybe these should be comments on the member variables?

alphan replied:
Good point, done.

mcy commented on sw/device/lib/dif/dif_usbdev.h:
```diff
@@ -0,0 +1,772 @@
+ * Configuration for initializing a USB device.
+ */
+typedef struct dif_usbdev_config {
+  mmio_region_t base_addr;
+  dif_usbdev_toggle_t differential_rx;
+  dif_usbdev_toggle_t differential_tx;
+  dif_usbdev_toggle_t single_bit_eop;
+  dif_usbdev_power_sense_override_t power_sense_override;
```
Could you provide comments for each of these member variables?

alphan replied:
Sure, done.

mcy commented on sw/device/lib/dif/dif_usbdev.h:
```diff
@@ -0,0 +1,772 @@
+  /**
+   * Indicates that the call succeeded and the entire packet payload is read.
+   */
+  kDifUsbdevRxPacketReadBytesResultOK = 0,
```
Rather than doing `= 0`, Maybe these should reference the constants in the "main" error enum?

alphan replied:
I agree.

mcy commented on sw/device/lib/dif/dif_usbdev.h:
```diff
@@ -0,0 +1,772 @@
+ * iterate over all available interrupts.
+ */
+typedef enum dif_usbdev_irq {
+  kDifUsbdevIrqFirst,
+  kDifUsbdevIrqPktReceived = kDifUsbdevIrqFirst,
+  kDifUsbdevIrqPktSent,
+  kDifUsbdevIrqDisconnected,
+  kDifUsbdevIrqHostLost,
+  kDifUsbdevIrqLinkReset,
+  kDifUsbdevIrqLinkSuspend,
+  kDifUsbdevIrqLinkResume,
+  kDifUsbdevIrqAvEmpty,
+  kDifUsbdevIrqRxFull,
+  kDifUsbdevIrqAvOverflow,
+  kDifUsbdevIrqLinkInError,
+  kDifUsbdevIrqRxCrcError,
+  kDifUsbdevIrqRxPidError,
+  kDifUsbdevIrqRxBitstuffError,
+  kDifUsbdevIrqFrame,
+  kDifUsbdevIrqConnected,
+  kDifUsbdevIrqLast = kDifUsbdevIrqConnected
```
Again, comments on all these.

alphan replied:
Done.

mcy commented on sw/device/lib/dif/dif_usbdev.h:
```diff
@@ -0,0 +1,772 @@
+ * iterate over all available interrupts.
+ */
+typedef enum dif_usbdev_irq {
+  kDifUsbdevIrqFirst,
+  kDifUsbdevIrqPktReceived = kDifUsbdevIrqFirst,
+  kDifUsbdevIrqPktSent,
+  kDifUsbdevIrqDisconnected,
+  kDifUsbdevIrqHostLost,
+  kDifUsbdevIrqLinkReset,
+  kDifUsbdevIrqLinkSuspend,
+  kDifUsbdevIrqLinkResume,
+  kDifUsbdevIrqAvEmpty,
+  kDifUsbdevIrqRxFull,
+  kDifUsbdevIrqAvOverflow,
+  kDifUsbdevIrqLinkInError,
+  kDifUsbdevIrqRxCrcError,
+  kDifUsbdevIrqRxPidError,
+  kDifUsbdevIrqRxBitstuffError,
+  kDifUsbdevIrqFrame,
+  kDifUsbdevIrqConnected,
+  kDifUsbdevIrqLast = kDifUsbdevIrqConnected
```
Other DIFs currently seem to mark `kLast` and friends as `\internal` in the doc comment. Consider doing this and removing the reference to them in this enum's overall type comment.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,1061 @@
+ * Note: This uses the bit indices of `USBDEV_IN_SENT` register for the sake
+ * of conciseness because other endpoint registers use the same layout.
+ */
+#define ENDPOINT_HW_INFO_ENTRY(N)                                 \
+  [N] = {.config_in_reg_offset = USBDEV_CONFIGIN##N##_REG_OFFSET, \
+         .bit_index = USBDEV_IN_SENT_SENT##N}
```
This is unpleasant, but I agree that it's a better algernative than risking getting the constants wrong.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,1061 @@
+
+#undef ENDPOINT_HW_INFO_ENTRY
+
+/**
+ * Mapping from `dif_usbdev_irq_t` to bit indices in interrupt registers.
+ */
+static const uint8_t kIrqEnumToBitIndex[] = {
+        [kDifUsbdevIrqPktReceived] = USBDEV_INTR_COMMON_PKT_RECEIVED,
+        [kDifUsbdevIrqPktSent] = USBDEV_INTR_COMMON_PKT_SENT,
+        [kDifUsbdevIrqDisconnected] = USBDEV_INTR_COMMON_DISCONNECTED,
+        [kDifUsbdevIrqHostLost] = USBDEV_INTR_COMMON_HOST_LOST,
+        [kDifUsbdevIrqLinkReset] = USBDEV_INTR_COMMON_LINK_RESET,
+        [kDifUsbdevIrqLinkSuspend] = USBDEV_INTR_COMMON_LINK_SUSPEND,
+        [kDifUsbdevIrqLinkResume] = USBDEV_INTR_COMMON_LINK_RESUME,
+        [kDifUsbdevIrqAvEmpty] = USBDEV_INTR_COMMON_AV_EMPTY,
+        [kDifUsbdevIrqRxFull] = USBDEV_INTR_COMMON_RX_FULL,
+        [kDifUsbdevIrqAvOverflow] = USBDEV_INTR_COMMON_AV_OVERFLOW,
+        [kDifUsbdevIrqLinkInError] = USBDEV_INTR_COMMON_LINK_IN_ERR,
+        [kDifUsbdevIrqRxCrcError] = USBDEV_INTR_COMMON_RX_CRC_ERR,
+        [kDifUsbdevIrqRxPidError] = USBDEV_INTR_COMMON_RX_PID_ERR,
+        [kDifUsbdevIrqRxBitstuffError] = USBDEV_INTR_COMMON_RX_BITSTUFF_ERR,
+        [kDifUsbdevIrqFrame] = USBDEV_INTR_COMMON_FRAME,
+        [kDifUsbdevIrqConnected] = USBDEV_INTR_COMMON_CONNECTED,
+};
```
Most other DIFs opt for a helper function with a `switch`. Is there a reason this approach is better in this case?

alphan replied:
I think separating validation and mapping of irqs resulted in simpler call sites in this case.

mcy replied:
I'm not gonna force you to change it, but I would consider switching to a function with a switch statement in it just to be consistent with the rest of the DIFs.

Is there a reason you really want to do an enum bounds check instead of catching it in a switch?

lenary replied:
I can see how in this case, doing it with an array makes it simpler.

Let's leave it like this and come back to it. USB has way more interrupts than quite a few other DIFs, and I'm not sure we fully know if we're going to move to a mask-based public interface for interrupts anyway.

mdhayter replied:
We could also consider changing the IP to lump all the error conditions into a single interrupt line with a separate error status?

Since the hardware does not match anything existing, we could certainly use driver writer's experience to change things.

alphan replied:
@mdhayter If I understand correctly, it would simplify the hw interface of usbdev but wouldn't really affect this part since callers would still be interested in individual bits.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,1061 @@
+ * @param entry A buffer entry.
+ * @return `true` if the operation was successful, `false` otherwise.
+ */
+static bool buffer_pool_add(dif_usbdev_buffer_pool_t *pool, uint8_t entry) {
```
Nit: since this is supposed to behave like a stack, maybe `push` and `pop` are more descriptive names? `remove` especially suggests removing a specific index.

alphan replied:
I went back and forth on this before using `add` and `remove`, but my rationale was: "This is a buffer pool, which happens to be implemented as a stack, so it feels more natural to have `add` and `remove` as opposed to `push` and `pop`." Please let me know what you think.

mcy replied:
Fair enough. I would still consider another name than `remove`, such as `take`. I'll leave it to your judgement.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,1061 @@
+ * @param entry A buffer entry.
+ * @return `true` if the operation was successful, `false` otherwise.
+ */
+static bool buffer_pool_add(dif_usbdev_buffer_pool_t *pool, uint8_t entry) {
```
Consider returning an error type instead? Though it seems you don't actually check the return value of this function anywhere...

alphan replied:
I think returning a `bool` is fine. I didn't check the return values of these functions because these are not expected to fail but bugs and attacks can always prove me wrong. Done.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,1061 @@
+ */
+static bool is_valid_power_sense_override(
+    dif_usbdev_power_sense_override_t val) {
+  return val == kDifUsbdevPowerSenseOverrideDisabled ||
+         val == kDifUsbdevPowerSenseOverridePresent ||
+         val == kDifUsbdevPowerSenseOverrideNotPresent;
```
For something like this, I would personally have written it as a switch, which feels more consise... but up to you.

alphan replied:
> but up to you.

I try to avoid fallthrough in switch statements whenever I can :)

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,1061 @@
+                                USBDEV_USBSTAT_AV_FULL) &&
+         !buffer_pool_is_empty(&usbdev->buffer_pool)) {
+    uint8_t buffer_entry;
+    buffer_pool_remove(&usbdev->buffer_pool, &buffer_entry);
```
This returns an error, yet it doesn't get acknowledged... this looks like a bug, but if it's not, consider at least writing `(void)` to make the fact that it's dropped on the ground explicit. (Applies throughout.)

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,1061 @@
+      (dif_usbdev_read_buffer_op_state_t){.buffer_entry = buffer_entry,
+                                          .remaining_bytes = packet->len,
+                                          .offset = 0};
+
```
Add a trailing comma to be consistent with the above.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,1061 @@
+  uint32_t config_in_val;
+  uint8_t buffer_entry;
+  uint8_t buffer_len;
+
```
Nit: bring the three predeclared variables down one line, so they're part of the `switch` stanza that initializes them.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,1061 @@
+    return kDifUsbdevTxPacketWriteBytesResultError;
+  }
+
+  // HW details
```
Drop this comment or elaborate it.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,1061 @@
+      USBDEV_CONFIGIN0_BUFFER0_OFFSET);
+
+  // Check the status of the packet
+  if (mmio_region_get_bit32(usbdev->base_addr, config_in_reg_offset,
```
Would it make sense to reorganize this control flow to avoid re-reading this particular register multiple times?

alphan replied:
Done.

mcy commented on sw/device/lib/dif/dif_usbdev.c:
```diff
@@ -0,0 +1,1061 @@
+      (bitfield_field32_t){.mask = USBDEV_USBCTRL_DEVICE_ADDRESS_MASK,
+                           .index = USBDEV_USBCTRL_DEVICE_ADDRESS_OFFSET,
+                           .value = addr});
+
```
Add trailing comma to help clang-format format this.

alphan commented:
Thanks for your comments @mcy!

- Added checks for return values of `buffer_pool_add/remove()`.
- Added comments.
- Improved some MMIO accesses.

mcy commented:
Alphan: 100% optional, but could you try adding `warn_unused_result` to declarations in this change? I want to see how it feels to use (it currently emits warnings, not errors).

alphan commented:
> Alphan: 100% optional, but could you try adding `warn_unused_result` to declarations in this change? I want to see how it feels to use (it currently emits warnings, not errors).

I think #2604 is illustrative enough. If you don't mind, I would like to wait until we reach a consensus there before updating this PR.

That being said, I added the attribute to all function declarations in this change (including the static ones in the c-file) and didn't get any warnings. I commented out some code just to see how it looks. Also, adding this to `void` functions results in ```error: 'warn_unused_result' attribute ignored [-Werror=attributes]``` as expected.

mcy commented:
NIce! Thanks for trying it out Alphan.

alphan commented:
Rebased and addressed the fallout from recent `mmio` changes. Will merge when all checks are done.

alphan commented:
Also updated #1601.
